rules/test/lldb/breakpoint.py

```diff
@@ -0,0 +1,81 @@
+# This program is a mechanism to dump a breakpoint spec from a file to a
+# `TestResult json dictionary.
+#
+# Design principals:
+# - Keep the LLDB code simple but emit much information
+# - Use LLDB's public command line API as much as possible
+# - Testing assertion logic runs outside of LLDB
+
+import lldb
+import json
+
+# Usage:
+"""
+command script import --allow-reload ./breakpoint.py
+breakpoint set --file 'tests/ios/app/App/main.m' --line 6
+breakpoint command add --python-function breakpoint.breakpoint_info_fn  1
+"""
+
+
+class TestResult():
+    def __init__(self, resolved_count, compilation_unit_file, expression_result, status):
+        self.resolved_count = resolved_count
+        self.compilation_unit_file = compilation_unit_file
+        self.expression_result = expression_result
+        self.status = status
+
+    def to_json(self):
+        return {
+            "resolved_count": self.resolved_count,
+            "compilation_unit_file": self.compilation_unit_file,
+            "expression_result": self.expression_result,
+            "status": self.status
+        }
+
+
+# breakpoint command add --python-function breakpoint.breakpointfn  1
+def breakpoint_info_fn(frame, bp_loc, extra_args, internal_dict):
+    debugger = frame.GetThread().GetProcess().GetTarget().GetDebugger()
+    hit_count = bp_loc.GetHitCount()
+    is_resolved = bp_loc.IsResolved()
+
+    # Resolve key information about the breakpoint, starting with basics
+    br = bp_loc.GetBreakpoint()
+    print("GOT_BR", br)
+    resolved_count = br.GetNumResolvedLocations()
+    print("GOT_RESOLVED_COUNT", resolved_count)
+
+    addr = bp_loc.GetAddress()
+    comp_unit = addr.GetCompileUnit()
+    compilation_unit_file = comp_unit.GetFileSpec().GetFilename()
+    print("GOT_COMPILATION_UNIT_FILE", compilation_unit_file)
+
+    # TODO: This should actually handle if variable is None per the test_rule
+    variable_name = None
+    with open("test_spec.json", 'r') as test_spec:
+        test_spec_dict = json.load(test_spec)
+        print("TEST_SPEC", test_spec_dict)
+        variable_name = test_spec_dict["variable_name"]
+
+    variable_result = frame.FindVariable(variable_name)
```

Comment:
Probably out of scope for this PR but instead of passing a variable through the rule, it might be better to pass an lldb command and use the `lldb.SBCommandInterpreter()`'s `HandleCommand()` function to execute an arbitrary command. The reason is that looking up a variable in the current frame doesn't break as often as looking up for stuff in the heap or compiling an expression...

Reply:
> The reason is that looking up a variable in the current frame doesn't break as often as looking up for stuff in the heap or compiling an expression...

Here's what it does today, my intention was to exercise both expression parsing and dynamic type resolution _after_ it successfully finds the frame variable: 
- evaluate the variable as an expression ( https://github.com/bazel-ios/rules_ios/pull/420/files/99159cc546be9113d6aca53da86c4d8383cf2deb#diff-36b15984d8cc3d857f6009b4c7d3a1cce0149cc3f1fa88feb4efcdea9157c8c8R67 )
- invoke the "Object Description" / assert on dynamic type info 

>  it might be better to pass an lldb command and use the lldb.SBCommandInterpreter()'s HandleCommand() function to execute an arbitrary command.

Totally agree, originally I was thinking this route and passing in `po` and the likes directly from starlark. The complexity lies in writing assertions against arbitrary commands: e.g. parsing the stdout. We'd probably hook a matcher lib like `llvm-lit` or `FileCheck` to assert on `stdout`. The core testing driver might be able to be refactored to more easily allow this!

Reply:
> it might be better to pass an lldb command and use the lldb.SBCommandInterpreter()'s HandleCommand() function to execute an arbitrary command.

Ok I refactored the exit handling in the latest PR and have added a Bazel rule + example of this: https://github.com/bazel-ios/rules_ios/pull/420/files#diff-36b15984d8cc3d857f6009b4c7d3a1cce0149cc3f1fa88feb4efcdea9157c8c8R110

Right now it just checks to see if the command has succeeded. We can refactor output assertions to do that but it'd likely be a follow up as this PR is getting very large. Probably we will keep the existing test case to instrument the frame at a lower level than the interpreter